### PR TITLE
ExaBGP integration

### PR DESCRIPTION
**Pull request by twirtgen**
Hi,

This PR is intended to add a new ExaBGP daemon to easily inject custom routes into any "real" BGP daemon.
To make this PR work I had to modify several things in IPmininet :

-  ExaBGP binds its IPv6 socket to an existing IPv6 address attached to the interface of the node. However, the IPv6 Duplicate Address Detection doesn't finished yet, and hence the daemon crashes when it starts. Now, the node wait until the DAD has finished before starting any daemon in the node
- On recent version of the BGP protocol, when establishing an eBGP sessions, import and export filters must be defined (RFC 8212). 
I had to define default import/export filters for the `ebgp_session` function. In the meantime, I observed a bug on the definition of access-list. IPmininet was defining, in exactly the same way, an access-list for both IPv4 and IPv6. However, access-lists are different according to the prefix family the daemon is processing.  So now I made a clear distinction between IPv4 and IPv6 route-maps
- I added prefix-list feature for filtering purposes

### CODE REVIEW

**Comment by jadinm** [comment]
Can one of the admins verify this patch?

File: ipmininet/router/config/utils.py [hunk 1]
```diff
@@ -36,3 +36,12 @@ def ip_statement(ip: Union[int, str, IPv6Address, IPv4Address]):
     if not isinstance(ip, int):
         ip = ip_interface(str(ip)).version
     return 'ipv6' if ip == 6 else 'ip'
+
+
+def family_translate(family: str):
```
Comment (jadinm) [inline ipmininet/router/config/utils.py:50]: Maybe it would be better to move that to a property of AddressFamily
This would remove the need of importing the method to the template

File: ipmininet/router/config/bgp.py [hunk 2]
```diff
@@ -306,8 +299,9 @@ def filters_to_match_cond(self,
                               filter_list: Sequence[Union[AccessList,
                                                           CommunityList,
                                                           PrefixList]],
-                              family):
+                              family=None):
```
Comment (jadinm) [inline ipmininet/router/config/bgp.py:302]: Why setting it to None if you only want either 'ipv4' or 'ipv6' ?

File: ipmininet/router/config/zebra.py [hunk 3]
```diff
@@ -11,6 +12,15 @@
 PERMIT = 'permit'
 
 
+def get_family(prefix: Union[IPv4Network, IPv6Network]) -> Optional[str]:
```
Comment (jadinm) [inline ipmininet/router/config/zebra.py:15]: The prefix can also be a string to convert to a prefix I think
`Union[str, IPv4Network, IPv6Network]`

The check would be `ip_network(str(prefix)).version == 4`, like ip_statement actually

File: ipmininet/router/config/zebra.py [hunk 4]
```diff
@@ -112,121 +122,156 @@ def __init__(self, name: Optional[str] = None, action=PERMIT,
...
+        return AccessListEntry
+
+
 class RouteMapMatchCond:
     """
     A class representing a RouteMap matching condition
     """
 
-    def __init__(self, cond_type: str, condition, family: Optional[str] = None):
+    def __init__(self, cond_type: str, condition, family=None):
```
Comment (jadinm) [inline ipmininet/router/config/zebra.py:256]: Why did you remove the type ?

**Review by jadinm — COMMENTED** [review]
Here is the first part of my review. I haven't had time to go through everything yet.
I will update it later on

File: ipmininet/router/config/zebra.py (hunk 4 above)
Comment (twirtgen) [reply ipmininet/router/config/zebra.py]: Oops, this this was due to an oversight in a local refactor. BTW changed in dca805acddbe19863aea03c4ed488785ca0b650e

File: ipmininet/router/config/zebra.py (hunk 3 above)
Comment (twirtgen) [reply ipmininet/router/config/zebra.py]: Changed: dca805acddbe19863aea03c4ed488785ca0b650e

File: ipmininet/router/config/bgp.py (hunk 2 above)
Comment (twirtgen) [reply ipmininet/router/config/bgp.py]: Oops, this this was due to an oversight in a local refactor. Changed in dca805acddbe19863aea03c4ed488785ca0b650e

File: ipmininet/router/config/utils.py (hunk 1 above)
Comment (twirtgen) [reply ipmininet/router/config/utils.py]: Changed: 1ea53a59706895d8a723d18302eedcdf4ddd2e5f

File: ipmininet/router/config/bgp.py [hunk 5]
```diff
@@ -445,20 +441,11 @@ def build_access_list(self) -> List[AccessList]:
...
+        return [] if node_access_lists is None else node_access_lists
 
     def build_prefix_list(self):
         node_prefix_lists = self._node.get('bgp_prefix_lists')
-        prefix_lists = []
-        if node_prefix_lists is not None:
-            for pfx_list in node_prefix_lists:
-                prefix_lists.append(PrefixList(name=pfx_list.name, entries=pfx_list.entries, family=pfx_list.family))
-        return prefix_lists
+        return [] if node_prefix_lists is None else node_prefix_lists
```
[the remark this answers is not in the document]
Comment (twirtgen) [reply ipmininet/router/config/bgp.py]: Changed: 69d91db6f7babd50a7fdb3e1b815b8e4d6833609

File: ipmininet/router/config/bgp.py [hunk 6]
```diff
@@ -445,20 +441,11 @@ def build_access_list(self) -> List[AccessList]:
         :return:
         """
         node_access_lists = self._node.get('bgp_access_lists')
-        access_lists = []
-        if node_access_lists is not None:
-            for acl_entries in node_access_lists:
-                access_lists.append(AccessList(name=acl_entries.name,
-                                               entries=acl_entries.entries))
-        return access_lists
+        return [] if node_access_lists is None else node_access_lists
```
[the remark this answers is not in the document]
Comment (twirtgen) [reply ipmininet/router/config/bgp.py]: Changed: 69d91db6f7babd50a7fdb3e1b815b8e4d6833609

File: ipmininet/router/config/exabgp.py [hunk 7]
```diff
@@ -246,16 +280,33 @@ def dry_run(self) -> str:
...
+
+                    * api.cli = 'false'
+
+                    * tcp.delay = 2 # waits at most 2 minutes
+        :param address_families: the routes to inject for both IPv4 and IPv6 unicast AFI.
+        :param passive: Tells to ExaBGP to not send active BGP Open messages. The daemon
+                        waits until the remote peer sends first the Open message to start
                         the BGP session. Its default value is set to True.
-        :return:
+        :return: The default configuration of this ExaBGP daemon
```
Comment (jadinm) [inline ipmininet/router/config/exabgp.py:309]: It does not really returns, only modifies

File: ipmininet/router/config/bgp.py [hunk 8]
```diff
@@ -488,7 +493,7 @@ class Peer:
     def __init__(self, base: 'Router', node: str, v6=False):
         """:param base: The base router that has this peer
         :param node: The actual peer"""
-        self.peer, other = self._find_peer_address(base, node, v6=v6)
+        self.peer, other, self.local_addr = self._find_peer_address(base, node, v6=v6)
```
Comment (jadinm) [inline ipmininet/router/config/bgp.py]: This is not always true. If BGP peers are not directly connected, the ```_find_peer_address``` does not find the local interface that is the closest to the peer. For that, you need to store it in the starting interface ```prio_queue``` and return it in the end

File: ipmininet/router/config/bgp.py [hunk 9]
```diff
@@ -507,7 +512,7 @@ def __init__(self, base: 'Router', node: str, v6=False):
 
     @staticmethod
     def _find_peer_address(base: 'Router', peer: str, v6=False) \
-            -> Tuple[Optional[str], Optional['Router']]:
+            -> Tuple[Optional[str], Optional['Router'], Optional[str]]:
         """Return the IP address that base should try to contact to establish
         a peering"""
```
Comment (jadinm) [inline ipmininet/router/config/bgp.py:551]: I think that it would be better to update the documentation of this method

**Review by jadinm — COMMENTED** [review]
Thanks for the update :)
Here is the last set of my comments before we can start to run the tests on jenkins

File: ipmininet/examples/exabgp_prefix_injector.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+                                   ExaList(["%d:%d" % (j, k) for j, k in zip(rnd_list(24, True), rnd_list(24, True))]))
+        med = BGPAttribute("med", random.randint(1, __MAX_UINT32_))
+        origin = BGPAttribute("origin", random.choice(["igp", "egp", "incomplete"]))
+
+        my_routes.append(BGPRoute(ip_network, [next_hop, origin, med, as_path, communities]))
+
+    return my_routes
+
+
+class ExaBGPTopoInjectPrefixes(IPTopo):
```
Comment (jadinm) [inline ipmininet/examples/exabgp_prefix_injector.py:66]: Also, I forgot but could you add a test that checks your example as well in ```tests/test_hexabgp.py``` ?

File: ipmininet/router/config/bgp.py (hunk 8 above)
Comment (twirtgen) [reply ipmininet/router/config/bgp.py]: I think that was changed in 53a7cd5224f7b281dabbd3fefafa6ba906b01749. But anyway, this is modified in 9af8167c1c77825d5e1d14ca1e89318c727975e8.  I had to slightly change the priority queue format because `IPIntf` does not override the `__lt__` method

File: ipmininet/router/config/bgp.py (hunk 9 above)
Comment (twirtgen) [reply ipmininet/router/config/bgp.py]: Changed: 4af2eb9c67192d505d51b10d1ca6f22936272221

File: ipmininet/router/config/exabgp.py (hunk 7 above)
Comment (twirtgen) [reply ipmininet/router/config/exabgp.py]: Changed: 4af2eb9c67192d505d51b10d1ca6f22936272221

File: ipmininet/router/__router.py [hunk 11]
```diff
@@ -124,6 +124,18 @@ def start(self):
         # Set relevant sysctls
         for opt, val in self.nconfig.sysctl:
             self._old_sysctl[opt] = self._set_sysctl(opt, val)
+
+        # wait until NDP has finished to check each IPv6 addresses assigned
+        # to the interface of the node.
+        # The command lists addresses failing duplicate address detection (IPv6)
+        # If any, it waits until all addresses has been checked
+        lg.debug(self._processes.node.name, 'Checking for any "tentative" addresses')
+        while a := self._processes.call("ip addr show tentative"):
```
[the remark this answers is not in the document]
Comment (twirtgen) [reply ipmininet/router/__router.py]: Changed: 4af2eb9c67192d505d51b10d1ca6f22936272221

File: docs/daemons.rst [hunk 12]
```diff
@@ -132,6 +132,123 @@ The following code shows how to use all these abstractions:
...
+                                                                            ExaList(
+                                                                                ["2914:480", "2914:413", "2914:4621"]))]),
+            BGPRoute(ip_network("1:5ee:bad:c0de::/64"), [BGPAttribute("next-hop", "self"),
+                                                         BGPAttribute("as-path", ExaList([1, 89, 42, 5])),
+                                                         BGPAttribute("origin", "igp")])
+        ]
+    }
+
+
+    class ExaBGPTopo(IPTopo):
```
[the remark this answers is not in the document]
Comment (twirtgen) [reply docs/daemons.rst]: Changed in 4af2eb9c67192d505d51b10d1ca6f22936272221. Modified to fit to the new API in 54536988c7c16870a8d09a16ae3179b7f2ff7bc3

File: ipmininet/examples/exabgp_prefix_injector.py (hunk 10 above)
Comment (twirtgen) [reply ipmininet/examples/exabgp_prefix_injector.py:114]: I refactor the tests from 2c5c17d70344122330a1cd74ad55de669f554fad to fb6a877e58a430e3ee1ab416dfb0dc691a14fd3f

**Comment by twirtgen** [comment]
Concerning 59007bca9d3fd681266ba5281e20929dff29c6d2, I think the old DFS did not correctly update the `visited` set (bgp.py `_find_peer_address`). I now put the string representation of the IPIntf in the visited set, instead of the object directly

File: ipmininet/install/install.py [hunk 13]
```diff
@@ -170,6 +171,27 @@ def install_openr(output_dir: str, may_fail=False):
...
 
 
+def install_exabgp(output_dir: str, may_fail=False):
+    git_url = "https://github.com/Exa-Networks/exabgp.git"
+    exabgp_src_folder = "exabgp-%s-src" % ExaBGPVersion
+    exabgp_path_src_dir = os.path.join(output_dir, exabgp_src_folder)
+    exabgp_self_executable = os.path.join(output_dir, "exabgp")
+    final_link = "/usr/sbin/exabgp"
+
+    sh("git clone {url} -o {src_dir}".format(url=git_url, src_dir=exabgp_src_folder),
```
Comment (jadinm) [inline ipmininet/install/install.py:181]:
```suggestion
    sh("git clone {url} {src_dir}".format(url=git_url, src_dir=exabgp_src_folder)
```

If you want the next command to work, you need to set the clone directory and not the origin with "-o" :)

**Comment by jadinm** [comment]
Jenkins, run tests.

File: ipmininet/install/install.py (hunk 13 above)
Comment (twirtgen) [reply ipmininet/install/install.py]: Yes indeed, it is more than likely that it was a bad copy and paste on my part. Anyway, it is changed a30819b8627b6c014fbe5949e7cf782dbbb8c413

**Comment by twirtgen** [comment]
Jenkins, run tests.

**Comment by twirtgen** [comment]
Jenkins, run tests.

**Comment by twirtgen** [comment]
Jenkins, run tests.

**Comment by twirtgen** [comment]
Jenkins, run tests.

**Comment by twirtgen** [comment]
Jenkins, run tests.

**Comment by twirtgen** [comment]
Jenkins, run tests.

**Comment by twirtgen** [comment]
Jenkins, run tests.

**Comment by twirtgen** [comment]
Jenkins, run tests.

**Comment by twirtgen** [comment]
Jenkins, run tests.

**Comment by twirtgen** [comment]
Jenkins, run tests.
